subprojects/gst-plugins-good/gst/rtp/gstrtph266pay.c

```diff
@@ -0,0 +1,792 @@
+      gst_adapter_clear (rtph266pay->adapter);
+      break;
+    default:
+      break;
+  }
+
+  ret = GST_ELEMENT_CLASS (parent_class)->change_state (element, transition);
+
+  // Chain up to the parent class first to avoid a race condition
+  // Check commit df724c410b02b82bd7db893d24e8572a06c2fcb1
+  switch (transition) {
+    case GST_STATE_CHANGE_PAUSED_TO_READY:
+      gst_rtp_h266_pay_clear_xps_cache (rtph266pay);
+      break;
+    default:
+      break;
+  }
+
+  return ret;
+}
+
+static gboolean
+gst_rtp_h266_pay_set_caps (GstRTPBasePayload * rtpbasepay, GstCaps * caps)
+{
+  GstRtpH266Pay *rtph266pay = GST_RTP_H266_PAY (rtpbasepay);
+  const gchar *alignment_str;
+  GstStructure *s;
+
+  gst_rtp_base_payload_set_options (rtpbasepay, "video", TRUE, "H265", 90000);
+
+  s = gst_caps_get_structure (caps, 0);
+  g_assert (s);
+  alignment_str = gst_structure_get_string (s, "alignment");
+  if (alignment_str) {
+    if (g_str_equal (alignment_str, "au")) {
+      rtph266pay->alignment = ALIGNMENT_AU;
+    } else if (g_str_equal (alignment_str, "nal")) {
+      rtph266pay->alignment = ALIGNMENT_NAL;
+    } else {
+      rtph266pay->alignment = ALIGNMENT_UNKOWN;
+    }
+  }
+
+  return TRUE;
+}
+
+static GstFlowReturn
+gst_rtp_h266_pay_handle_buffer (GstRTPBasePayload * rtpbasepay,
+    GstBuffer * buffer)
+{
+  GstRtpH266Pay *rtph266pay = GST_RTP_H266_PAY (rtpbasepay);
+  GstAdapter *adapter = rtph266pay->adapter;
+  gsize adapter_size;
+
+  GST_DEBUG_OBJECT (rtph266pay, "New buffer %" GST_PTR_FORMAT, buffer);
+  gst_adapter_push (adapter, buffer);
+
+  // Advance until the first start code to skip heading zeroes
+  adapter_size = gst_adapter_available (adapter);
+  gssize next_nalu_start =
+      gst_adapter_masked_scan_uint32 (adapter, NALU_SC_MSK, NALU_SC_VAL, 0,
+      adapter_size);
+  if (next_nalu_start < 0) {
+    GST_WARNING_OBJECT (rtph266pay, "Not NALU found");
+    gst_adapter_flush (adapter, adapter_size);
+    return GST_FLOW_OK;
+  }
+  gst_adapter_flush (adapter, next_nalu_start);
+
+  // For each NALU
+  while ((adapter_size = gst_adapter_available (adapter)) > NALU_SC_LEN) {
+    // Find next start code, skipping the actual one
+    gssize next_nalu_start =
+        gst_adapter_masked_scan_uint32 (adapter, NALU_SC_MSK, NALU_SC_VAL,
+        NALU_SC_LEN, adapter_size - NALU_SC_LEN);
+
+    // If no start code found, the length of the NALU is the remaining size
+    gssize nalu_len = (next_nalu_start < 0) ? adapter_size : next_nalu_start;
+
+    if (nalu_len <= NALU_SC_LEN) {
+      GST_WARNING_OBJECT (rtph266pay, "NALU too small %ld, skipping it",
+          nalu_len);
+      gst_adapter_flush (adapter, nalu_len);
+      continue;
+    }
+
+    // Remove the start code
+    nalu_len -= NALU_SC_LEN;    // Here, it'll always be > 0
+    gst_adapter_flush (adapter, NALU_SC_LEN);
+
+    GstBuffer *nalu_buf = gst_adapter_take_buffer (adapter, nalu_len);
+    GST_BUFFER_PTS (nalu_buf) = gst_adapter_prev_pts (adapter, NULL);
```

Comment:
Why not `GST_BUFFER_PTS (nalu_buf) = GST_BUFFER_PTS (buffer)`? Similar for DTS.

Reply:
`gst_adapter_take_buffer ()` ends calling `gst_buffer_copy_region()` [which doesn't copy the timestamps when the offset != 0](https://github.com/fluendo/gstreamer/blob/main/subprojects/gstreamer/gst/gstbuffer.c#L587-L603). The offset will always be greater than zero here (even for `alignment=au`) as we always remove the NAL prefix before.

The `gstrtph265pay` does the same. Maybe it's worth adding a little commentary here :thinking: ...

Reply:
This does not explain why `GST_BUFFER_PTS (nalu_buf) = GST_BUFFER_PTS (buffer)` would not be equivalent. But, now I see that gstrtph265pay is doing the same.